### DISCUSSION TITLE
charon not closing the TCTI resource manager device (/dev/tpmrm0)

### DISCUSSION THREAD

**Discussion by sdiw** [strongswan/strongswan#846] [Q&A] [answered]
I am using the latest version of strongswan 5.9.4 and forming a tunnel mode connection with left auth as eap-tls and using a TPM for the private key. I am able to form the tunnel with no problem but the issue is that, upon loading the credentials, charon keeps the device file /dev/tpmrm0 open and as a result this disallows any other tool/process from using the TPM during that time.

```
/root # lsof | grep tpm

/root # swanctl -q
no files found matching '/usr/local/etc/swanctl/conf.d/*.conf'
loaded certificate from '/usr/local/etc/swanctl/x509/ak_rsa_cert.pem'
loaded certificate from '/usr/local/etc/swanctl/x509ca/int1.pem'
loaded certificate from '/usr/local/etc/swanctl/x509ca/int2.pem'
loaded certificate from '/usr/local/etc/swanctl/x509ca/int3.pem'
loaded certificate from '/usr/local/etc/swanctl/x509ca/network2_remote_primary.pem'
loaded key token_ak_pitunnel from token [keyid: 6cd9e5ade70731198df867ab6ad07a29e30f39c3]
no authorities found, 0 unloaded
no pools found, 0 unloaded
loaded connection 'pitunnel'
successfully loaded 1 connections, 0 unloaded

/root # lsof | grep tpm
2564	/usr/local/libexec/ipsec/charon	/dev/tpmrm0

/root # tpm2_getcap hanldes-permanent
ERROR:tcti:src/tss2-tcti/tcti-device.c:478:Tss2_Tcti_Device_Init() Failed to open specified TCTI device file /dev/tpmrm0: Resource busy
WARNING:tcti:src/tss2-tcti/tctildr.c:79:tcti_from_init() TCTI init for function 0x7fab1e5fc0 failed with a000a
WARNING:tcti:src/tss2-tcti/tctildr.c:109:tcti_from_info() Could not initialize TCTI named: tcti-device
ERROR:tcti:src/tss2-tcti/tctildr-dl.c:150:tcti_from_file() Could not initialize TCTI file: libtss2-tcti-device.so.0
ERROR:tcti:src/tss2-tcti/tcti-device.c:478:Tss2_Tcti_Device_Init() Failed to open specified TCTI device file /dev/tpm0: Resource busy
WARNING:tcti:src/tss2-tcti/tctildr.c:79:tcti_from_init() TCTI init for function 0x7fab1e5fc0 failed with a000a
WARNING:tcti:src/tss2-tcti/tctildr.c:109:tcti_from_info() Could not initialize TCTI named: tcti-device
ERROR:tcti:src/tss2-tcti/tctildr-dl.c:150:tcti_from_file() Could not initialize TCTI file: libtss2-tcti-device.so.0

```
Is this intended? I assume charon needs access to the TPM only during [re]keying.
Why does it have to keep the device open at all times?

**Comment by strongX509** [comment, marked as answer]
Hi, the strongSwan `libtpmtss` library by default uses the `/dev/tpmrm0` device which has a built-in resource manager that allows multiple connections to the TPM 2.0 device. On my various Linux installations I don't have any problems for multiple connections to access the TPM. It's very strange that `tpm2_getcap` reports `TCTI device file /dev/tpmrm0: Resource busy`. Usually the strongSwan `charon` daemon has two active private TPM 2.0 AK keys for signatures and I can still use the `tpm2-tools` at the same time without blocking. 

What versions of the `tpm2-tss` library and the `tpm2-tools` are you using?

**Reply by sdiw** [reply]
Hi Andreas Steffen, thanks for your response. I am using the versions `tpm2-tss-2.4.0` and `tpm2-tools-4.2`.
I am seeing the same problem when loading two private TPM 2.0 keys or a private TPM 2.0  key and a certificate from the nvram.

Can you please confirm if the device file `/dev/tpmrm0` remains open while the `charon` daemon has active private TPM 2.0 keys?
